### PR TITLE
Fixed physical GPU ID to logical conversions

### PR DESCRIPTION
**Pull request by jhh67**
The CUDA and HIP libraries refer to GPUs by unique IDs that are consecutive integers starting at zero. When GPUs are shared by co-locales the IDs of the GPUs used by a particular locale may not start with zero and they may not be consecutive, so they cannot be used as the sublocale ID or to index into arrays. Previous modifications to support co-locales introduced "physical" (global) IDs vs "logical" (relative) IDs, but it wasn't a very clean implementation and recent changes to memory management broke it. This PR cleans up the implementation, makes it more obvious what is going on, and adds various sanity checks.

Tested against `test/gpu/native/multiLocale` both without co-locales and with `CHPL_RT_LOCALES_ON_NODE=2` for both `CHPL_GPU=amd` and `CHPL_GPU=nvidia`. 

Resolves https://github.com/Cray/chapel-private/issues/6817.
Resolves https://github.com/chapel-lang/chapel/issues/26104.

### CODE REVIEW

**Review by DanilaFe — APPROVED** [review]

File: runtime/src/gpu/amd/gpu-amd.c [hunk 1]
```diff
@@ -396,7 +418,8 @@ void chpl_gpu_impl_mem_free(void* memAlloc) {
     // see note in chpl_gpu_mem_free
     hipPointerAttribute_t res;
     ROCM_CALL(hipPointerGetAttributes(&res, memAlloc));
-    switch_context(res.device);
+    int dev_lid = dev_pid_to_lid(res.device);
+    switch_context(dev_lid);
```
Comment (jabraham17) [inline runtime/src/gpu/amd/gpu-amd.c:422]: `hipSetDevice` needs a physical id and you already have one here, but you have to roundtrip it to the logical id just to call `switch_context`. This pattern makes me think there should be 2 `switch_contexts`, one that takes a logical id and one that takes a physical id.

**Review by jabraham17 — COMMENTED** [review]
The code looks good to me, but I think you should add tests to the `test/gpu/native/multiLocale` directory for colocales, covering all the cases listed in the PR description.

You can probably just clone `test/gpu/native/multiLocale/onAllGpusOnAllLocales.chpl` to `test/gpu/native/multiLocale/colocales/onAllGpusOnAllLocales.chpl`, and then have tests for colocales with GPUs, to prevent regressions

**Comment by DanilaFe** [comment]
> The code looks good to me, but I think you should add tests to the test/gpu/native/multiLocale directory for colocales, covering all the cases listed in the PR description.

I think the idea is to test that entire directory with colocales. I'm looking into it, particularly since we probably want to only run such tests in systems that support rhem.

**Review by jabraham17 — APPROVED** [review]

File: runtime/src/gpu/amd/gpu-amd.c (hunk 1 above)
Comment (jhh67) [reply runtime/src/gpu/amd/gpu-amd.c:422]: I decided it wasn't worth the extra function. It's only a couple of array accesses.

**Pull request merged**